pwnlib/libcdb.py

```diff
@@ -645,10 +787,76 @@ def search_by_sha256(hex_encoded_id, unstrip=True):
         >>> hex(ELF(filename).symbols.read)
         '0xeef40'
     """
-    return search_by_hash(hex_encoded_id, 'sha256', unstrip)
+    return search_by_hash(hex_encoded_id, 'sha256', unstrip, offline, online)
+
+def _get_libc_symbols(path):
+    """
+    Given symbol file path, the content format is as follows:
+        ...
+        puts 000xxxxx
+        read 0000xxxx
+        write 0000xxxx
+        ...
+
+    Arguments:
+        path(str):
+            symbol file path.
+
+    Returns:
+        a dictionary of the form: { name : address }, or :const:`{}`
+    """
+
+    syms = {}
 
+    with open(path, "r") as fd:
+        for x in fd.read().strip().split("\n"):
+            name, addr = x.split(" ")
+            syms[name] = int(addr, 16)
 
+    return syms
 
+def _pack_libs_info(path, libs_id, libs_url, syms):
+    """ 
+    The JSON format is the same as libc.rip, and the "download_url" field is by default an empty string, 
+    as it's not required in offline mode.
+    """
+
+    info = {}
+
+    info["id"] = libs_id
+    info["libs_url"] = libs_url
+    info["download_url"] = ""
+
+    info["buildid"] = _get_build_id(path)
+    info["md5"] = md5filehex(path)
+    info["sha1"] = sha1filehex(path)
+    info["sha256"] = sha256filehex(path)
+
+    sym_list = [
+        "__libc_start_main_ret", "dup2", "open", "printf", "puts", "read", "system", "str_bin_sh"
+    ]
+
+    info["symbols"] = {}
+    for name in sym_list:
+        info["symbols"][name] = hex(syms[name])
+
+    return info
+
+def _get_build_id(path):
+    """
+    Arguments:
+        path(str):
+            elf file path.
+
+    Returns:
+        hexdigital build id, or :const:`None`
+    """
+
+    with open(path, "rb") as fd:
+        section = ELFFile(fd).get_section_by_name('.note.gnu.build-id')
```

Comment:
In fact, it will be much faster, and there will be a lot of unnecessary output when using ELF().

use `ELF(path, checksec=False).buildid`:

<img width="1773" alt="iShot_2023-11-23_12 45 02" src="https://github.com/Gallopsled/pwntools/assets/46804083/f4871ea0-6f45-43a9-a3f3-bdd3cc50991a">

use `ELFFile(fd).get_section_by_name('.note.gnu.build-id')`:

<img width="965" alt="iShot_2023-11-23_12 45 41" src="https://github.com/Gallopsled/pwntools/assets/46804083/c7043db3-6757-4fd2-bd7a-af0c483dee9b">

Reply:
I'm testing on the following code

```py
from elftools.elf.elffile import ELFFile

def _get_elf_buildid(path):
    """ Some ELF files lack the `.note.gnu.build-id` section. """

    build_id = None

    with open(path, "rb") as fd:
        section = ELFFile(fd).get_section_by_name('.note.gnu.build-id')
    
        if section:
            build_id = section.data()

    if not build_id:
        log.debug("%s does not have a buildid", path)
        return None

    return binascii.b2a_hex(build_id).decode()

def _get_elf_buildid(path):
    """ Some ELF files lack the `.note.gnu.build-id` section. """

    build_id = ELF(path, checksec=False).buildid

    if not build_id:
        log.debug("%s does not have a buildid", path)
        return None

    return binascii.b2a_hex(build_id).decode()
```

Reply:
You can use `context.quiet` to silence the warnings (and they should be addressed anyway I think?).

Reply:
How to address the time consumption, Is there any method to reduce the runtime of ELF? Using ELFFile takes only about 1 second, while ELF takes around 17 seconds.

And can you take a look at this issue #2304?